lighthouse-core/lib/tracehouse/trace-processor.js

```diff
@@ -74,11 +132,33 @@ class TraceProcessor {
       }
     }
 
-    // sort by ts, if there's no ts difference sort by index
-    indices.sort((indexA, indexB) => {
-      const result = traceEvents[indexA].ts - traceEvents[indexB].ts;
-      return result ? result : indexA - indexB;
-    });
+    // Sort by ascending timestamp first.
+    indices.sort((indexA, indexB) => traceEvents[indexA].ts - traceEvents[indexB].ts);
+
+    // Now we find groups with equal timestamps and order them by their nesting structure.
+    for (let i = 0; i < indices.length - 1; i++) {
+      const ts = traceEvents[indices[i]].ts;
+      const tsGroupIndices = [indices[i]];
```

Comment:
while playing around with this code, I will say having `tsGroupIndices` being indexes into `traceEvents` but having to look up other indexes indirectly through`timestampSortedIndices` first ended up with me confusing myself a few times and having to step through the code to reassure myself about what was going on.

It definitely feels like the kind of code that is going to be frightening for other people to touch in the future, and/or when people make changes there will be subtle errors that will be hard to track down without reading the whole file :)

What about being consistent and passing in `i` here, so it always has to be an indirect lookup? Could maybe be made better with a little lookup function declared in `_sortTimestampEventGroup` or something?

Reply:
(the other option would be to actually do the `sorted` step first and then loop back over and sort events with the same ts in the newly sorted array, but not sure how that would look)

Reply:
yeah agreed it's confusing, will go with consistent lookup approach 👍